enhancements/network/openstack-dual-stack.md

```diff
@@ -0,0 +1,188 @@
+4. The customer will generate the manifests and configured MachineConfigs
+for the master and workers depending on the IPv6 addresses mode chosen for the Subnets
+in step 1. If SLAAC or Stateless was used the MachineConfigs has to contain `ip_addr_mode=eui64`
+and `method=auto` in the connection for the ethernet interface. While when using stateful
+it has to enable `ip=dhcp,dhcpv6` in the kernel arguments.
+5. The `uuid` from the network has to be added in the worker MachineSet alongside with the
+subnets that were pre-filled for the machine-api-provider-openstack to create one unique Port
+with both addresses configured for the Server.
+6. The CIDRs of the subnets have to added to the `cloud-provider-config.yaml`
+in the `address-sort-order` field to make sure IPv4 address will have higher priority in the
+Node addresses.
+7. The installer will validate the subnets and VIPs, add the security group to the api and ingress
+ports and link the floating IPs to those Ports. Also, create the additional IPv6 security group rules
+and configured the servers with dual-stack addresses and add the dual-stack allowed address pairs.
+8. In machine-config-operator, kubelet will be update to not be configured with node-ips.
+9. cloud-provider-openStack will filter the node addresses that are meant to be specified in the Status of the Node
+by looking for a tag that was applied in the main Ports of the Node.
+
+### API Extensions
+
+The `machinesSubnet` field, which is available in the `install-config.yaml` under
+`platform` and `openstack` will be deprecated in favor of a new field named `machinesSubnets`
+to persist both IPv4 and IPv6 provided Subnets.
+
+For installation on OpenStack with dual-stack kubelet will no longer rely on the node-ip to get the node
+addresses, given that when using external cloud provider, the node-ip currently has no support for dual-stack, this [enhancement](https://github.com/kubernetes/enhancements/pull/3706) describes it. Instead,
+the node addresses will be provided directly by the cloud provider openstack, which can rely on a tag to filter them. This [issue](https://github.com/kubernetes/cloud-provider-openstack/issues/2096) details the idea behind it.
+
+### Implementation Details/Notes/Constraints
+
+The idea behind using bring your own network instead of basic IPI installation is to allow support
+for a wider variety of ways on how to configure the Network and Subnet. As an example, the Ports have
+to be pre-created by the user because when using SLAAC or Stateless the specification of fixed-ips during
+Port creation is not allowed. UPI can also be an option in future releases.
+
+### Risks and Mitigations
+
+We need to validate if the proposal for not setting the node-ip will be a robust enough
+solution that we want to maintain, specially around upgrades. If not robust, we can use the dual-stack support for
+node-ip that is being [developed upstream](https://github.com/kubernetes/kubernetes/pull/116305).
```

Comment:
Robust is the wrong word here: robustness isn't really a worry; it should be more robust that the current solution, which relies on heuristics which we have already broken. The solution is also very simple, with one very significant caveat:

The issue is that we haven't given any thought at all to upgrades. We have no idea what the complexity of that problem is.

Reply:
I know there is a strong resistance against `--node-ip` functionality *but if* it turns out that the concept of Tags makes the upgrade&conversion too complex will we be able to fallback to using `--node-ip` with 2 IP addresses? I assume for 4.14 you want to have vanilla dual-stack installations covered and then for 4.15 an upgrade&conversion story; that would mean in the worst-case in 4.15 you return to `--node-ip` and pass dual-stack IPs.

Does it even stick together? Or is introducing those Tags now will make us stick with them forever?

Reply:
I think we have to. I also don't think we've started the tags work, so I strongly suspect we'll be using --node-ip.